source/common/upstream/cluster_factory_impl.cc

```diff
@@ -0,0 +1,153 @@
+#include "common/upstream/cluster_factory_impl.h"
+
+#include "common/http/utility.h"
+#include "common/network/address_impl.h"
+#include "common/network/resolver_impl.h"
+#include "common/network/socket_option_factory.h"
+#include "common/upstream/health_checker_impl.h"
+
+#include "server/transport_socket_config_impl.h"
+
+namespace Envoy {
+namespace Upstream {
+
+namespace {
+
+Stats::ScopePtr generateStatsScope(const envoy::api::v2::Cluster& config, Stats::Store& stats) {
+  return stats.createScope(fmt::format(
+      "cluster.{}.", config.alt_stat_name().empty() ? config.name() : config.alt_stat_name()));
+}
+
+} // namespace
+
+ClusterSharedPtr ClusterFactoryImplBase::create(
+    const envoy::api::v2::Cluster& cluster, ClusterManager& cluster_manager, Stats::Store& stats,
+    ThreadLocal::Instance& tls, Network::DnsResolverSharedPtr dns_resolver,
+    Ssl::ContextManager& ssl_context_manager, Runtime::Loader& runtime,
+    Runtime::RandomGenerator& random, Event::Dispatcher& dispatcher,
+    AccessLog::AccessLogManager& log_manager, const LocalInfo::LocalInfo& local_info,
+    Server::Admin& admin, Singleton::Manager& singleton_manager,
+    Outlier::EventLoggerSharedPtr outlier_event_logger, bool added_via_api, Api::Api& api) {
+
+  std::string cluster_type;
+
+  if (!cluster.has_cluster_type()) {
+    switch (cluster.type()) {
+    case envoy::api::v2::Cluster::STATIC:
+      cluster_type = Extensions::Clusters::ClusterTypes::get().Static;
+      break;
+    case envoy::api::v2::Cluster::STRICT_DNS:
+      cluster_type = Extensions::Clusters::ClusterTypes::get().StrictDns;
+      break;
+    case envoy::api::v2::Cluster::LOGICAL_DNS:
+      cluster_type = Extensions::Clusters::ClusterTypes::get().LogicalDns;
+      break;
+    case envoy::api::v2::Cluster::ORIGINAL_DST:
+      cluster_type = Extensions::Clusters::ClusterTypes::get().OriginalDst;
+      break;
+    case envoy::api::v2::Cluster::EDS:
+      // We map SDS to EDS, since EDS provides backwards compatibility with SDS.
+      cluster_type = Extensions::Clusters::ClusterTypes::get().Eds;
+      break;
+    default:
+      NOT_REACHED_GCOVR_EXCL_LINE;
+    }
+  } else {
+    cluster_type = cluster.cluster_type().name();
+  }
+  ClusterFactory* factory = Registry::FactoryRegistry<ClusterFactory>::getFactory(cluster_type);
+
+  if (factory == nullptr) {
+    throw EnvoyException(fmt::format(
+        "Didn't find a registered cluster factory implementation for name: '{}'", cluster_type));
```

Comment:
Please make sure you have test coverage for this using EXPECT_THROW_WITH_MESSAGE